src/pkg/retention/policy/action/performer.go

```diff
@@ -50,32 +54,71 @@ type retainAction struct {
 // Perform the action
 func (ra *retainAction) Perform(candidates []*art.Candidate) (results []*art.Result, err error) {
 	retained := make(map[string]bool)
+	immutable := make(map[string]bool)
+	retainedShare := make(map[string]bool)
+	immutableShare := make(map[string]bool)
 	for _, c := range candidates {
-		retained[c.Hash()] = true
+		retained[c.NameHash()] = true
+		retainedShare[c.Hash()] = true
+	}
+
+	for _, c := range ra.all {
+		if _, ok := retainedShare[c.Hash()]; ok {
+			continue
+		}
+		if isImmutable(c) {
+			immutable[c.NameHash()] = true
+			immutableShare[c.Hash()] = true
+		}
 	}
 
 	// start to delete
 	if len(ra.all) > 0 {
```

Comment:
strongly suggest adding the immutable property into the candidate when we retrieving them. Execute the immutable rules at that time.